components/admin_console/admin_definition.jsx

```diff
@@ -648,6 +654,306 @@ export default {
+                            help_text_default: 'When true, Mattermost periodically synchronizes users from AD/LDAP. When false, user attributes are updated in Mattermost during user login.',
+                        },
+                        {
+                            type: Constants.SettingsTypes.TYPE_TEXT,
+                            key: 'LdapServer',
+                            label: 'admin.ldap.serverTitle',
+                            label_default: 'AD/LDAP Server:',
+                            help_text: 'admin.ldap.serverDesc',
+                            help_text_default: 'The domain or IP address of AD/LDAP server.',
+                            placeholder: 'admin.ldap.serverEx',
+                            placeholder_default: 'E.g.: "10.0.0.23"',
+                            needs_or: [['Enable', true], ['EnableSync', true]],
+                        },
+                        {
+                            type: Constants.SettingsTypes.TYPE_TEXT,
+                            key: 'LdapPort',
+                            label: 'admin.ldap.portTitle',
+                            label_default: 'AD/LDAP Port:',
+                            help_text: 'admin.ldap.portDesc',
+                            help_text_default: 'The port Mattermost will use to connect to the AD/LDAP server. Default is 389.',
+                            placeholder: 'admin.ldap.portEx',
+                            placeholder_default: 'E.g.: "389"',
+                            needs_or: [['Enable', true], ['EnableSync', true]],
+                        },
+                        {
+                            type: Constants.SettingsTypes.TYPE_DROPDOWN,
+                            key: 'ConnectionSecurity',
+                            label: 'admin.connectionSecurityTitle',
+                            label_default: 'Connection Security:',
+                            help_text: DefinitionConstants.CONNECTION_SECURITY_HELP_TEXT_LDAP,
+                            options: [
+                                {
+                                    value: '',
+                                    display_name: 'admin.connectionSecurityNone',
+                                    display_name_default: 'None',
+                                },
+                                {
+                                    value: 'TLS',
+                                    display_name: 'admin.connectionSecurityTls',
+                                    display_name_default: 'TLS (Recommended)',
+                                },
+                                {
+                                    value: 'STARTTLS',
+                                    display_name: 'admin.connectionSecurityStart',
+                                    display_name_default: 'STARTTLS',
+                                },
+                            ],
+                            needs_or: [['Enable', true], ['EnableSync', true]],
```

Comment:
Yes, we can start the conversation, and I'm more than happy to debate it :)

About the need of extending for each edge case, I think we have to provide the tools to be quite adaptative. For example, the "Custom" widget allow you to have a widget fully customized for your own form, don't need to build a new widget for each new type.

I'm thinking about some kind of "life-cycle-functions" something like pre-save, post-save, pre-change, post-change, or whatever we need to achieve the behaviors of the current admin console.

This approach have some important advantages, one of the is that we only have to deal with a small code base which manage show each widget is shown, so if you have bug, you solve it in the component, and everything is fixed. Once it is done, it could be easy to extract it and replicate it into ReactNative, only need to implement the widgets.

As side effect (not a coincidence at all) we will be able to know what is in the admin menu and index the text easily.

Another interesting side effect is, this allow the plugins to have a more powerful set of widgets to configure in the admin panel.